#Author: JP Gallagher
#Last updated: 07/Aug/2014
#The purpose of this script is to look for gene conversion within a set of
#alignments.  This conservative method will look at SNPs between A and D
#diploids and determine if those are shared by the appropriate homoeologs as 
#generated by polyCat. The final result should be a list of alignments that
#maintain A- and D-specific SNPs without conversion to the other gene copy.

#Note that input is a NEXUS file where species have names in the format of 
#"A1_Wagad" or "AD1_Maxxa". The species code must come first, followed by an
#underscore, followed by whatever specific information on the accession is
#necessary

#set working directory while writing/debugging this script
#setwd("~/Projects/GeneConversionMethods")

#this package provides DNA alignment handling tools
require(ape)

#Get a list of all the alignment files
aln.files <- list.files(pattern = "Gorai.*")
good.alignments.A <- vector()
good.alignments.D <- vector()

#Then we write a for loop for reading in alignments, calculating A and D
#SNPs for each alignment, 
for(i in 1:length(aln.files)){
  #read in NEXUS file
  file <- aln.files[i]
  print(file)
  nexus <- read.nexus.data(file)
  
  #get the names of the accessions
  taxa <- NULL
  taxa <- names(nexus)
  
  #get all the A's and all the D's
  A.species <- grep("^A[12]",taxa)
  D.species <- grep("^D5",taxa)

  #remake alignment with only the parental diploids
  diploids.only.aln <- nexus[c(A.species,D.species)]
  A.diploids.only.aln <- nexus[A.species]
  D.diploids.only.aln <- nexus[D.species]
  
  #find the SNPs that separate the diploids
  diploid.SNPs <- seg.sites(as.DNAbin(diploids.only.aln))
  A.SNPs <- seg.sites(as.DNAbin(A.diploids.only.aln))
  #D.SNPs <- seg.sites(as.DNAbin(D.diploids.only.aln))
  allSNPs <- c(diploid.SNPs,A.SNPs)#had to get rid of D.SNPs because it is only one sequence now, add back in if you want to use more than just D5
  #remove all those that are duplicated
  AvD.shared.SNPs <- allSNPs[!(duplicated(allSNPs) | duplicated(allSNPs,fromLast=TRUE))]
  
  #get At and Dt accession names
  At.acc <- grep("At",taxa)
  Dt.acc <- grep("Dt",taxa)

  #make nexus with A species and At sequences and D and Dt species
  AplusAt.aln <- nexus[c(A.species,At.acc)]
  DplusDt.aln <- nexus[c(D.species,Dt.acc)]

  #get segregating sites
  #seg.sites ignores sequences with Ns at certain site
  At.SNPs <- seg.sites(as.DNAbin(AplusAt.aln))
  Dt.SNPs <- seg.sites(as.DNAbin(DplusDt.aln))
  
  #Is the segregating SNP present between A and At or D and Dt?
  if(!any(duplicated(c(At.SNPs,AvD.shared.SNPs)))){
    good.alignments.A <- append(good.alignments.A,file)
  }
  if(!any(duplicated(c(Dt.SNPs,AvD.shared.SNPs)))){
    good.alignments.D <- append(good.alignments.D,file)
  }
}

write(good.alignments.A,"good.alignments.A.txt")
write(good.alignments.D,"good.alignments.D.txt")
